### PR TITLE
rgw: exclude logging of request payer for 403 requests

### PR DESCRIPTION
**Pull request by clwluvw**
As per AWS doc (https://docs.aws.amazon.com/AmazonS3/latest/userguide/RequesterPaysBuckets.html#ChargeDetails), request payer should not incur charges for 403 requests.

Fixes: https://tracker.ceph.com/issues/65473

### CODE REVIEW

File: src/rgw/rgw_log.cc [hunk 1]
```diff
@@ -200,25 +200,21 @@ static void log_usage(req_state *s, const string& op_name)
...
+  string bucket_name = s->bucket_name;
 
   if (!bucket_name.empty()) {
-    bucket_name = s->bucket_name;
     user = to_string(s->bucket_owner.id);
+
     if (!rgw::sal::Bucket::empty(s->bucket.get()) &&
-	s->bucket->get_info().requester_pays) {
+        s->bucket->get_info().requester_pays &&
+        s->err.http_ret != 403) {
```
Comment (cbodley) [inline src/rgw/rgw_log.cc:212]: ok, aws docs have specific handling for 403s; could you please add a code comment that describes this?

File: src/rgw/rgw_log.cc [hunk 2]
```diff
@@ -200,25 +200,21 @@ static void log_usage(req_state *s, const string& op_name)
...
       payer = s->user->get_id().to_str();
     }
-  } else {
-    user = to_string(s->owner.id);
   }
 
-  bool error = s->err.is_err();
-  if (error && s->err.http_ret == 404) {
+  if (rgw::sal::Bucket::empty(s->bucket.get())) {
```
Comment (cbodley) [inline src/rgw/rgw_log.cc:217]: why did this need to change?

**Review by cbodley — COMMENTED** [review]
👍

File: src/rgw/rgw_log.cc (hunk 2 above)
Comment (clwluvw) [reply src/rgw/rgw_log.cc:217]: It's a bit confusing - I guess a 404 HTTP status code could also mean an object was not found or any other related to the action and doesn't necessarily mean the bucket is not found and if we have the bucket in `s->bucket` it means the bucket is valid. Am I mistaken?

Comment (cbodley) [reply src/rgw/rgw_log.cc:217]: i don't know why this needs to do anything other than `bucket_name = s->bucket_name;` above, really

but this seems unrelated to the request payer change. it should be a separate commit, at least

Comment (clwluvw) [reply src/rgw/rgw_log.cc:217]: I guess it makes sense to have it not flood the usagelog with invalid buckets (as each bucket would result in a unique `rgw_usage_log_entry`).
I'll remove it from this PR and propose a new PR so we can discuss it there. It's irrelevant from this PR and ticket as you said.

Comment (cbodley) [reply src/rgw/rgw_log.cc:217]: ok. i didn't even realize this was for the usage log, i thought it was for the ops log

File: src/rgw/rgw_log.cc [hunk 3]
```diff
@@ -200,21 +200,24 @@ static void log_usage(req_state *s, const string& op_name)
...
+
+    // As per https://docs.aws.amazon.com/AmazonS3/latest/userguide/RequesterPaysBuckets.html#ChargeDetails
+    // If the bucket has requester pays enabled,
+    // and the requerster includes x-amz-request-payer in the header,
+    // and the status code isn't 403,
+    // then the requester is the payer.
     if (!rgw::sal::Bucket::empty(s->bucket.get()) &&
-	s->bucket->get_info().requester_pays) {
+        s->bucket->get_info().requester_pays &&
+        s->get_request_payer() &&
```
Comment (clwluvw) [inline src/rgw/rgw_log.cc:217]: I found that we're not checking for the header existence of `x-amz-request-payer` also, so I changed the commit in this regard.

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-04-24_12:59:55-rgw-wip-cbodley-testing-distro-default-smithi/ with rerun https://pulpito.ceph.com/cbodley-2024-04-24_20:45:10-rgw-wip-cbodley-testing-distro-default-smithi/

**Pull request merged**